### ISSUE TITLE
New eval/exec/compile + import bypass

### ISSUE THREAD

**Issue by discriminating** [mache102/puso#12]
![lol](https://cdn.discordapp.com/attachments/953021220662706196/1097566888646688849/image.png)

I found another bypass. By using getattr and getting the builtin function eval, you can set it to your own variable which will transform into a function, bypassing your string compare. 

How to fix?
On line 121 in puso.py, you can change the line `if 'eval(' in line or 'exec(' in line or 'compile(' in line:` to `if 'eval' in line or 'exec' in line or 'compile' in line:` to avoid this. This is because you are searching for the function call, rather than the name.

Code for bypass:
```py
encrypt = lambda s: ''.join(chr((ord(c) + 3 - 32) % 95 + 32) if ' ' <= c <= '~' else c for c in s)

encrypted = encrypt("__import__('os').system('dir')")
print(f"bypassed = getattr(__builtins__, 'eval'); decrypt = lambda s: bypassed(''.join(chr((ord(c) - 3 - 32) % 95 + 32) if ' ' <= c <= '~' else c for c in s)); decrypt('{encrypted}');print;")```

**Comment by Remi-Avec-Un-I** [comment]
There is a better way to forbide import, this by completly changing what does python when importing a module, you can do that using the "builtins" modules : 
```
import builtins
def _block_import(name, globals=None, locals=None, fromlist=None, level=0):
        raise SyntaxError("importing modules is not supported")
def run():
    builtins.__import__ = _block_import
```
Now, instead of importing a module, it will always call the block import function

**Comment by discriminating** [comment]
Good point, that would work better

**Comment by Remi-Avec-Un-I** [comment]
And it doesn't require to disable any other function like eval, exec etc...

**Comment by mybearworld** [comment]
That's not going to work, the changes will just apply to the `puso.builtins` module. Even if they applied to the program importing the module, you could just revert all changes to the module by reimporting `puso.builtins`.

**Comment by Remi-Avec-Un-I** [comment]
> 

If you want to revert puso.builtins.__import__ to normal, you would need to import builtins, and you can't

**Comment by mache102** [comment]
Interesting bypass
I updated the exec() checks and implemented the builtins override. 
Does builtin also allow for overriding other function (range, exec, etc)? 
I suppose I could also make eval, exec, compile do nothing

**Comment by discriminating** [comment]
Yes, I'm pretty sure builtin allows control over all default python functions that can be called with no imports.

**Comment by discriminating** [comment]
Follow up: Can be bypassed once again by string concatenation

![lmao](https://cdn.discordapp.com/attachments/1092864599256215642/1098082712360857672/image.png)

**Comment by Remi-Avec-Un-I** [comment]
> Interesting bypass I updated the exec() checks and implemented the builtins override. Does builtin also allow for overriding other function (range, exec, etc)? I suppose I could also make eval, exec, compile do nothing

```print(dir(builtins))``` You can override all of python native function (see : https://docs.python.org/3/library/functions.html) And I didn't understand the bypass found by @discriminating

**Comment by discriminating** [comment]
> > Interesting bypass I updated the exec() checks and implemented the builtins override. Does builtin also allow for overriding other function (range, exec, etc)? I suppose I could also make eval, exec, compile do nothing
> 
> `print(dir(builtins))` You can override all of python native function (see : https://docs.python.org/3/library/functions.html) And I didn't understand the bypass found by @discriminating

My bypass works because because of a flaw in your fix. You see, as @mybearworld pointed out, overriding the import function using builtins is local to the module, and does not replicate to the file imported in.

See here:
![sefwerdfed](https://cdn.discordapp.com/attachments/1092864599256215642/1098247034659684403/image.png)

Okay, so, what if we implement your fix in the local file?
![oopsies](https://cdn.discordapp.com/attachments/1092864599256215642/1098248204161663147/image.png)

Ooookaayyy, so what if we fix that?
![wtf](https://user-images.githubusercontent.com/99683656/233101180-8f83a2e0-fdc6-4c2d-a4bd-ac502481ff46.png)

My best guess is to allow a one-liner at the top (that is string compared) that will rewrite the builtins, though there may be a better way

**Comment by Remi-Avec-Un-I** [comment]
But isn't it the case because you specify a custom_fp ? I don't get what it's used for

**Comment by mache102** [comment]
I think I'll pause the updates for import blocking for now, it's killing my sanity
